### DISCUSSION TITLE
Support for SCSI Stream Commands (for tape emulation)

### DISCUSSION THREAD

**Discussion by Kars-de-Jong** [ZuluSCSI/ZuluSCSI-firmware#153] [Ideas]
Currently the tape emulation is not useable. Most importantly, reading a tape doesn't work because the command parameter block of several commands (for example the READ(6) command) gets quite a different meaning. See for instance this working draft: https://www.cs.cmu.edu/afs/club/usr/jhutz/project/Archives/scsi/ssc-r22.pdf 

In order to support this, some state needs to be kept which is tape-specific, such as the current position in the file (subsequent tape READ commands continue where the previous command ended, unlike disk READ commands which always specify the LBA).

The most important commands that need to work to at least be able to install something from tape (as far as I have seen from the debug log so far) are:

0x01 - REWIND - Should set the current position to 0.
0x08 - READ(6) - Should start reading at the current position, updating the current position.
0x11 - SPACE - Basically a seek command for tapes, with special features (go to file mark).

There's also special handling required for End-Of-Tape condition.

How to handle the concept of file marks is to be handled with a raw image file is unclear to me. See for a bit of background info this page for instance: https://www.gnu.org/software/tar/manual/html_section/Many.html. Maybe we could support the .T64 format, that seems fairly simple (see https://vice-emu.sourceforge.io/vice_17.html#SEC391). A simple start could be to handle it similar to how multiple CD images can be used now, resetting the image index to 0 on rewind and indicating end-of-tape when the last image is fully read.

My particular use case is emulation of an OMTI 5400 (see http://www.bitsavers.org/pdf/sms/OMTI_5x00.pdf). It predates the SCSI-standard, in some cases the commands are not entirely compatible but this can be handled by the already existing OMTI quirk.

I'm guessing this should really be done upstream in the SCSI2SD repository by the way.

I'm willing to help (I can code), but I'd like to discuss the architecture a bit first.

**Comment by aperezbios** [comment]
@Kars-de-Jong thanks for reaching out. This has been on our list of things to eventually get working. There is at least one incomplete implementation that I'm aware of, which might be a good starting point. @reinauer wrote said code. I'll ping him.

**Comment by Kars-de-Jong** [comment]
I managed to hack something together which allowed me to install an operating system from tape.
It turned out the OMTI quirk wasn't quite there yet. I will create two branches for this work, one for the OMTI updates and one for the tape support.

**Comment by Kars-de-Jong** [comment]
I created PRs for the OMTI updates and for rudimentary tape support.
